File: typescript/src/days/day03.ts
// Import NodeJS helpers
import fs from 'fs';

// Import types
import InputFile from '../types/InputFile';

// ----------------- Helpers ----------------- //

const splitRucksack = (rucksack: string): [string, string] => {
  const mid = rucksack.length / 2;
  return [rucksack.slice(0, mid), rucksack.slice(mid)];
};

const rucksackOverlap = (rucksacks: [string, string]): string => {
  const ruckMap = new Map<string, boolean>();

  [...rucksacks[0]].forEach((char) => ruckMap.set(char, true));

  return [...rucksacks[1]].find((char) => ruckMap.has(char));
};

const splitGroups = (rucksacks: string[]): string[][] => {
  const groups = [];

  for (let i = 0; i < rucksacks.length - 3; i += 3) {
    groups.push([
      rucksacks[i],
      rucksacks[i + 1],
      rucksacks[i + 2],
    ]);
  }

  return groups;
};

const groupOverlap = (group: string[]): string => {
  const ruckMap = new Map<string, boolean>();

  [...group[0]].forEach((char) => ruckMap.set(char, false));

  [...group[1]].forEach((char) => (
    ruckMap.has(char)
      ? ruckMap.set(char, true)
      : null
  ));

  return [...group[2]].find((char) => ruckMap.has(char) && ruckMap.get(char));
};

const priority = (char: string) => {
  if (!char) return 0;

  const ord = char.charCodeAt(0);

  if (ord > 96 && ord < 123) return ord - 96;
  if (ord > 64 && ord < 91) return ord - 38;
};

// ------------- Main Functions -------------- //

/**
 * Day 3 Pt1 of Advent of Code 2022!
 * Given a list of pairs of rucksacks, what items do the pairs share?
 * @author Benedikt Arnarsson
 */
const pt1 = async (filename: InputFile) => {
  const inputFile = await fs.promises.readFile(filename);

  return inputFile
    .toString()
    .split('\n')
    .map(splitRucksack)
    .map(rucksackOverlap)
    .map(priority)
    .reduce((sum, nxt) => sum + nxt, 0);
};

/**
 * Day 3 Pt2 of Advent of Code 2022!
 * Given groups-of-three, what item does each group share?
 * @author Benedikt Arnarsson
 */
const pt2 = async (filename: InputFile) => {
  const inputFile = await fs.promises.readFile(filename);

  const rucksacks = inputFile.toString().split('\n');
  const groups = splitGroups(rucksacks);

  return groups
    .map(groupOverlap)
    .map(priority)
    .reduce((sum, nxt) => sum + nxt, 0);
};

const day03 = [pt1, pt2];

export default day03;

File: python/days/day03.py
"""Day 3 of AOC 2022: Rucksack Reorganization"""

# -------------- Public Functions ------------- #

def day03_pt1(filename: str) -> int:
    """Day 3, part 1: Check for shared items in rucksack pairs."""
    with open(filename, 'r', encoding='utf-8') as input_file:
        rucksacks = input_file.read().split('\n')[:-1]

    shared_items = map(_rucksack_overlap, rucksacks)
    priorities = map(_get_priority, shared_items)

    return sum(priorities)

def day03_pt2(filename: str) -> int:
    """Day 3, part 2: Check group-of-three rucksack overlap"""
    with open(filename, 'r', encoding='utf-8') as input_file:
        rucksacks = input_file.read().split('\n')[:-1]

    num_groups = len(rucksacks)//3
    groups = [[rucksacks[3*i+j] for j in range(3)] for i in range(num_groups)]

    shared_items = map(_group_overlap, groups)
    priorities = map(_get_priority, shared_items)

    return sum(priorities)


# ----------------- Constants ----------------- #


# ------------- Private Functions ------------- #

def _rucksack_overlap(rucksack: str) -> str:
    mid = len(rucksack)//2
    [rucksack_1, rucksack_2] = rucksack[:mid], rucksack[mid:]
    # Inefficient but 'Pythonic'
    for char in rucksack_1:
        if char in rucksack_2:
            return char

    raise Exception('This should not happen...\nCheck ../inputs/02.txt')

def _group_overlap(group: list[str]) -> str:
    [elf1, elf2, elf3] = group

    for char in elf1:
        if char in elf2 and char in elf3:
            return char

    raise Exception('This should not happen...\nCheck ../inputs/02.txt')


def _get_priority(char: str) -> int:
    try:
        ascii_val = ord(char)
        if ascii_val in range(97, 123):
            return ascii_val - 96
        elif ascii_val in range(65, 91):
            return ascii_val - 38
        else:
            return 0
    except TypeError as exc:
        raise Exception('This should not happen...\nCheck ../inputs/02.txt') from exc

# ------------------- Script ------------------ #

if __name__ == '__main__':
    INPUT_FILENAME = '../inputs/03.txt'
    print(f'Day 3, pt1: {day03_pt1(INPUT_FILENAME)}')
    print(f'Day 3, pt2: {day03_pt2(INPUT_FILENAME)}')
